### ISSUE TITLE
Prefill INSTALLDIR at project upgrade with last used path

### ISSUE THREAD

**Issue by Grandik7** [oleg-shilo/wixsharp#1380]
Hi, 

While installing app using ManagedUI at InstallDir dialog there is always constant path - even when ugrade is installed and previous version was installed to some custom location. Is there any simple way how to obtain and preset this path?

Thanks & regards,
Michal

**Comment by Torchok19081986** [comment]
morning, AFAIK msi get Path for Update / Upgrade by GUID of product and perform uninstall and install after. You can get Installdir Path in Session . public property is INSTALLDIR, which owned by msi itself. IF you want to display it, You can try to create CA and get it from there .

Example Code , not tested . 

```csharp

[CustomAction]

public ActionResult GetInstallDir(Session session)
{
string installdir = session["INSTALLDIR"].ToString();

return ActionResult.Success;
}

```

One important Note : Startingpoint of this CA. I dont know, if CA belong to UI Sequence or not. 🙄🙄🙄

**Comment by Grandik7** [comment]
Hi Torchok, 
thank for your reaction.

To clarify - my project is setup as below:
![image](https://github.com/oleg-shilo/wixsharp/assets/38108481/99d68ae0-243f-47f7-bca1-aa24436202fa)
This path is always preset in UI as default InstallDir.

Test case: I've installed app in some other path. When installing update (for now letting InstallDir dialog as is - with default path), BeforeInstall event is fired two times: first time after installation starts - session contains required custom InstallDir path (therefore it is stored somewehre, that's good). Second time it fires with session containing real path where software is installed - default path, at the end of installation.

Question is, what event should I use to obtain custom path before InstallDir is displayed and set it accordingly. 
UILoaded and UIInitialized events have InstallDir property empty in SetupEventArgs.

Thanks & regards, 
Michal

**Comment by Torchok19081986** [comment]
thanks for more info. AFAIK in Event Beforeinstall you has another properties like IsUninstalling, IsBeingUpgraded. Try these in BeforeInstall. 
I think, you have to combine Before.- and AfterInstall in your specific case. And in Afterinstall you has direct access to Installdir. 
Which values MSI shows in Before and AfterInstall Event ? You can set messagebox to Show() Method in Testproject there.

**Comment by Grandik7** [comment]
In every event, there is SetupEventArgs argument containing:
![image](https://github.com/oleg-shilo/wixsharp/assets/38108481/f3bdb970-67b6-4322-90a9-ad59e5ed2e5a)
Unofortunatelly no event with desired path is fired before InstallDir dialogs show up ... Maybe there is need to use registers :/

**Comment by Torchok19081986** [comment]
Installdir Path is your standardpath, right ? you use ManagedUI , means you have full control to UI. Try to overwrite INSTALLDIR  Property in InstalldirDialog in Event Window_Load .

**Comment by Grandik7** [comment]
InstallDirDialog does not have any public property - I have to create custom dialog. 
ManagedProject.ManagedUI has InstallDirId property - this would not help as directory that user chose is not added in project dirs.

And still the biggest pitfall is, that I am not able to obtain directory of last installation early.

**Comment by oleg-shilo** [comment]
Let's have a fresh look at it. 

You need to find a way of retrieving the install dir of the previous installation. It's not an "MSI" task but a task for your product. You can use any sort of persistence for that. Let's assume you solved it (e.g. on the first run the product permanently saves its location/installdir in the registry). Now we can move on to the actual challenge.

If you are installing the product for the first time:
- Setup with or without UI
   1. check the registry for the installdir; detect that it is empty 
   2. run the msi session as usual

If you are installing the product on the machine where it was previously installed:
- Setup without UI
  1. As early as possible (before `LaunchCondition` action), check the registry for the install dir; read its value
  2. Push the install dir value to the current msi session:
      You might need to schedule this custom action for `Sequence.InstallExecuteSequence` as well. You will need to experiment
        ```C#
          new ManagedAction(Script.SetInstallDir,
                            Return.check,
                            When.Before,
                            Step.LaunchConditions,
                            Condition.NOT_Installed,
                            Sequence.InstallUISequence),
                    . . .
          [CustomAction]
          public static ActionResult SetInstallDir(Session session)
          {
              // set custom installdir
              // This event is fired before native MSI UI loaded 
              session["INSTALLDIR"] = Registry.CurrentUser
                                                    .OpenSubKey(@"SOFTWARE\MyProduct")
                                                    .GetValue("Path")
                                                    .ToString();
              return ActionResult.Success;
          }
    ```

- Setup with Managed UI
   1. In the InstallDir dialog, check the registry for the install dir; read its value
   2. Set the installdir value to the InstallDir text box:
       ```C#
        void InstallDirDialog_Load(object sender, EventArgs e)
        {
            banner.Image = Runtime.Session.GetResourceBitmap("WixSharpUI_Bmp_Banner");

            installDirProperty = Runtime.Session.Property("WixSharp_UI_INSTALLDIR");

            string installDirPropertyValue = Runtime.Session.Property(installDirProperty);

            if (installDirPropertyValue.IsEmpty())
            {
                // We are executed before any of the MSI actions are invoked so the INSTALLDIR (if set to absolute path)
                // is not resolved yet. So we need to do it manually
                installDir.Text = Runtime.Session.GetDirectoryPath(installDirProperty);
                
                var  oldInstalldir = Registry.CurrentUser
                                                    .OpenSubKey(@"SOFTWARE\MyProduct")
                                                    .GetValue("Path")?
                                                    .ToString();
                if (oldInstalldir.HasText())
                    installDir.Text = oldInstalldir;
            }
            else
            {
                //INSTALLDIR set either from the command line or by one of the early setup events (e.g. UILoaded)
                installDir.Text = installDirPropertyValue;
            }

            ResetLayout();
        }
       ```

**Comment by Grandik7** [comment]
Mr. Oleg, 

thank you for very helpful and clear comment!

Anyway, I have one more curious question: why 'retrieving the install dir of the previous installation' is not a msi task? When unninstalling app, installer obviously knows where app is installed. Therefore I assume this information is somewhere somehow stored. 

Thanks & regards,
Michal

**Comment by Grandik7** [comment]
Thank you Oleg, I am almost there. 

I am setting InstallDir in ManagedProject.UiLoaded event - therefore there is no need to create custom InstallDirDialog. 

I am trying to create registry key in Installer also - in ManagedProject.AfterInstall event. Only issue is, that installer runs under the Local System session, not current user session. Is there any way how to execute some code as current user in this event?

**Comment by oleg-shilo** [comment]
Yes there is. You can just call `project.UnelevateAfterInstallEvent()` before the build.
